### PR TITLE
Remove some preprocessor usage from the JIT interface.

### PR DESCRIPTION
**Pull request by pgavlin**
This fixes most of #4912.

- The usage of `USE_ASM_GC_WRITE_BARRIERS` has been removed. The type it guarded is
  now always defined.
- The JIT helpers that were guarded by `_TARGET_X86_` are now always defined.
- The intrinsic that was guarded by `_WIN64` is now always defined.

### CODE REVIEW

**Comment by pgavlin** [comment]
@russellhadley @jkotas @BruceForstall PTAL

**Comment by pgavlin** [comment]
cc @dotnet/jit-contrib

**Comment by jkotas** [comment]
LGTM

File: src/inc/corinfo.h [hunk 1]
```diff
@@ -663,7 +663,6 @@ enum CorInfoHelpFunc
     // Keep platform-specific helpers at the end so that the ids for the platform neutral helpers stay same accross platforms
     //
 
-#if defined(_TARGET_X86_) || defined(_HOST_X86_) // _HOST_X86_ is for altjit
                                     // NOGC_WRITE_BARRIERS JIT helper calls
```
Comment (BruceForstall) [inline src/inc/corinfo.h]: Even if you delete the #ifdefs, it seems like you should include some comments stating these are only used by X86 compilations. The comment immediately above this one also seems to now be obsolete.

File: src/inc/corjit.h [hunk 2]
```diff
@@ -301,7 +301,6 @@ enum CorJitFuncKind
     CORJIT_FUNC_FILTER         // a funclet associated with an EH filter
 };
 
-#if !defined(FEATURE_USE_ASM_GC_WRITE_BARRIERS) && defined(FEATURE_COUNT_GC_WRITE_BARRIERS)
 // We have a performance-investigation mode (defined by the FEATURE settings above) in which the
```
Comment (BruceForstall) [inline src/inc/corjit.h]: The comment below needs to be updated. You should probably mention the #defines above in the comment, since they are still used to build the VM and JIT.

File: src/inc/corinfo.h [hunk 3]
```diff
@@ -231,11 +231,11 @@ TODO: Talk about initializing strutures before use
...
-    0x7a6aa61a,
-    0x78b1,
-    0x4dfb,
-    { 0x9e, 0x6, 0x65, 0x5f, 0xb4, 0x77, 0x4d, 0x7f }
+SELECTANY const GUID JITEEVersionIdentifier = { /* 718c4238-2a85-45de-88ad-9b1fed806547 */
+    0x718c4238,
+    0x2a85,
+    0x45de,
+    { 0x88, 0xad, 0x9b, 0x1f, 0xed, 0x80, 0x65, 0x47 }
 };
```
Comment (BruceForstall) [inline src/inc/corinfo.h]: You have been making lots of changes to the JIT-EE interface. Can't you batch them all up, to make fewer of them? Changing the JIT-EE interface has heavy associated costs.

**Comment by BruceForstall** [comment]
:shipit:

File: src/inc/corjit.h (hunk 2 above)
Comment (pgavlin) [inline src/inc/corjit.h]: Done. We should consider enabling this mode with a JIT flag instead of preprocessor syumbols.

File: src/inc/corinfo.h (hunk 1 above)
Comment (pgavlin) [inline src/inc/corinfo.h]: Done.

File: src/inc/corinfo.h (hunk 3 above)
Comment (pgavlin) [inline src/inc/corinfo.h]: I've been trying to batch them up based on whether or not they can be performed in GitHub. I know that this creates some extra work w.r.t. SuperPMI collections, but I'd prefer to do as many of these changes in GH as possible.

Comment (BruceForstall) [inline src/inc/corinfo.h]: That is optimizing for the wrong thing.

Comment (pgavlin) [inline src/inc/corinfo.h]: I disagree. We should try to make changes in the open and to keep unrelated changes separate for clearer history as much as possible. If there are processes that are getting in the way of those goals, those processes should be improved.

Comment (BruceForstall) [inline src/inc/corinfo.h]: Of course processes can be improved. Also, you can submit unrelated changes separately, but at the same time, or batched.

**Comment by pgavlin** [comment]
OS X failure seems to be infrastructure-related: `14:29:11 /Users/dotnet-bot/j/workspace/dotnet_coreclr/master/checked_osx_prtest/Tools/Packaging.targets(814,5): error : Unable to find description for package NuGet3MinVersion [/Users/dotnet-bot/j/workspace/dotnet_coreclr/master/checked_osx_prtest/src/.nuget/Microsoft.NETCore.Runtime.CoreCLR/Microsoft.NETCore.Runtime.CoreCLR.pkgproj]`

cc @gkhanna79

**Comment by gkhanna79** [comment]
@ericstj Do you know what the above pkg compile error is about?

**Comment by ericstj** [comment]
@gkhanna79 this requirement was introduced in the latest buildtools.  See https://github.com/dotnet/corefx/commit/eae5aa4063bf5990292d0b753415f127ebe97755

**Comment by gkhanna79** [comment]
So, is the fix to pickup the description from the commit you shared? If so, @pgavlin can you please incorporate as part of your change?

**Comment by ericstj** [comment]
@joperezr should have done it with this fix: https://github.com/dotnet/coreclr/commit/c396cf15da1ddbf1fd7dcf256357adac44829679

**Comment by gkhanna79** [comment]
@dotnet-bot test OSX x64 Checked Build and Test please

**Comment by gkhanna79** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test please

**Comment by pgavlin** [comment]
The OS X job is going to fail due to a full disk. cc @mmitche

**Comment by mmitche** [comment]
Disk isn't full, it's still the semaphore issue seen in the past.

**Comment by pgavlin** [comment]
Any fix, or just retry?

**Comment by joperezr** [comment]
if this is not full perhaps just retry?

**Comment by joperezr** [comment]
@dotnet-bot test OSX x64 Checked Build and Test please

**Comment by pgavlin** [comment]
All of these jobs are now failing with `ERROR: [WS-CLEANUP] Cannot delete workspace: remote file operation failed: D:\j\workspace\checked_windo---f6dc6fe4 at hudson.remoting.Channel@3378981:Azure0520082559: hudson.remoting.ChannelClosedException: channel is already closed` cc @mmitche

**Comment by pgavlin** [comment]
The Ubuntu failure is in a PAL test.

**Comment by pgavlin** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by pgavlin** [comment]
The debug build failure is an infrastructural issue:

```
15:36:17 D:\j\workspace\debug_windows---17180f1b>powershell -Command "Add-Type -Assembly 'System.IO.Compression.FileSystem'; [System.IO.Compression.ZipFile]::CreateFromDirectory('.\bin\tests\Windows_NT.x64.Debug', '.\bin\tests\tests.zip')" 
15:36:20 ERROR: Build step failed with exception
15:36:20 java.lang.IllegalStateException: Invalid object ID 10 iota=11
15:36:20 	at hudson.remoting.ExportTable.diagnoseInvalidId(ExportTable.java:386)
15:36:20 	at hudson.remoting.ExportTable.get(ExportTable.java:330)
15:36:20 	at hudson.remoting.Channel.getExportedObject(Channel.java:633)
15:36:20 	at hudson.remoting.RemoteInvocationHandler$RPCRequest.perform(RemoteInvocationHandler.java:599)
15:36:20 	at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:583)
15:36:20 	at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:542)
15:36:20 	at hudson.remoting.UserRequest.perform(UserRequest.java:120)
15:36:20 	at hudson.remoting.UserRequest.perform(UserRequest.java:48)
15:36:20 	at hudson.remoting.Request$2.run(Request.java:326)
15:36:20 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
15:36:20 	at java.util.concurrent.FutureTask.run(Unknown Source)
15:36:20 	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
15:36:20 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
15:36:20 	at hudson.remoting.Engine$1$1.run(Engine.java:62)
15:36:20 	at java.lang.Thread.run(Unknown Source)
15:36:20 	at ......remote call to Azure0519083947(Native Method)
15:36:20 	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1416)
15:36:20 	at hudson.remoting.UserResponse.retrieve(UserRequest.java:220)
15:36:20 	at hudson.remoting.Channel.call(Channel.java:781)
15:36:20 	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:250)
15:36:20 	at com.sun.proxy.$Proxy74.join(Unknown Source)
15:36:20 	at hudson.Launcher$RemoteLauncher$ProcImpl.join(Launcher.java:991)
15:36:20 	at hudson.tasks.CommandInterpreter.join(CommandInterpreter.java:135)
15:36:20 	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:95)
15:36:20 	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:64)
15:36:20 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
15:36:20 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:782)
15:36:20 	at hudson.model.Build$BuildExecution.build(Build.java:205)
15:36:20 	at hudson.model.Build$BuildExecution.doRun(Build.java:162)
15:36:20 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:534)
15:36:20 	at hudson.model.Run.execute(Run.java:1738)
15:36:20 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
15:36:20 	at hudson.model.ResourceController.execute(ResourceController.java:98)
15:36:20 	at hudson.model.Executor.run(Executor.java:410)
15:36:20 Caused by: java.lang.Exception: Object was recently deallocated
15:36:20     #10 (ref.0) : object=null type=hudson.Launcher$RemoteLaunchCallable$1 interfaces=[hudson.Launcher$RemoteProcess]
```

All tests associated with the build passed.
